functions-python/validation_report_processor/src/main.py

```diff
@@ -0,0 +1,264 @@
+        raise Exception(f"Validation report {report_id} already exists.")
+
+    validation_report_entity = Validationreport(
+        id=report_id,
+        validator_version=version,
+        validated_at=validated_at,
+        html_report=html_report_url,
+        json_report=json_report_url,
+    )
+    entities.append(validation_report_entity)
+
+    dataset = get_dataset(dataset_stable_id, session)
+    for feature_name in json_report["summary"]["gtfsFeatures"]:
+        feature = get_feature(feature_name, session)
+        feature.validations.append(validation_report_entity)
+        entities.append(feature)
+    for notice in json_report["notices"]:
+        notice_entity = Notice(
+            dataset_id=dataset.id,
+            validation_report_id=report_id,
+            notice_code=notice["code"],
+            severity=notice["severity"],
+            total_notices=notice["totalNotices"],
+        )
+        entities.append(notice_entity)
+    return entities
+
+
+def create_validation_report_entities(feed_stable_id, dataset_stable_id):
+    """
+    Creates and stores entities based on a validation report.
+    This includes the validation report itself, related feature entities,
+    and any notices found within the report.
+
+    :param feed_stable_id: Stable ID of the feed
+    :param dataset_stable_id: Stable ID of the dataset
+    :return: Tuple List of all entities created (Validationreport, Feature, Notice) and status code
+    """
+    json_report_url = (
+        f"{FILES_ENDPOINT}/{feed_stable_id}/{dataset_stable_id}/report.json"
+    )
+    logging.info(f"Accessing JSON report at {json_report_url}.")
+    json_report, code = validate_json_report(json_report_url)
+    if code != 200:
+        return json_report, code
+
+    try:
+        validated_at, version = parse_json_report(json_report)
+    except Exception as error:
+        return str(error), 500
+
+    session = None
+    try:
+        session = start_db_session(os.getenv("FEEDS_DATABASE_URL"))
+        logging.info("Database session started.")
+
+        # Generate the database entities required for the report
+        try:
+            entities = generate_report_entities(
+                version,
+                validated_at,
+                json_report,
+                dataset_stable_id,
+                session,
+                feed_stable_id,
+            )
+        except Exception as error:
+            return str(error), 409  # Conflict if report already exists
+
+        # Commit the entities to the database
+        for entity in entities:
+            session.add(entity)
+        logging.info(f"Committing {len(entities)} entities to the database.")
+        session.commit()
+
+        logging.info("Entities committed successfully.")
+        return f"Created {len(entities)} entities.", 200
+    except Exception as error:
+        logging.error(f"Error creating validation report entities: {error}")
```

Comment:
do we need to roolback the DB session here?

Reply:
Absolutely -- added ✅